app/src/main/java/com/github/h3lp3rs/h3lp/messaging/Conversation.kt

```diff
@@ -16,26 +22,153 @@ class Conversation(
 ) {
     private val database = databaseOf(MESSAGES)
+    private var publicKey: PublicKey? = null
+    private val allMessages: HashMap<String, String> = HashMap()
+    private val keyAlias = "CONVERSATION_KEY_$conversationId"
+
+
+    init {
+
+        val keyStore = KeyStore.getInstance(ANDROID_KEY_STORE).apply { load(null) }
+        val alias = keyStore.aliases().toList()
+
+        if (!alias.contains(keyAlias)) {
+            val kpg =
+                KeyPairGenerator.getInstance(KeyProperties.KEY_ALGORITHM_RSA, ANDROID_KEY_STORE)
+
+            kpg.initialize(
+                KeyGenParameterSpec.Builder(
+                    keyAlias, KeyProperties.PURPOSE_ENCRYPT
+                            or KeyProperties.PURPOSE_DECRYPT
+                )
+                    .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_RSA_OAEP)
+                    .setDigests(KeyProperties.DIGEST_SHA1)
+                    .build()
+            )
+
+            val kp = kpg.generateKeyPair()
+
+            // Send public key to the database
+            val encodedPublicKey = Base64.encodeToString(kp.public.encoded, Base64.DEFAULT)
+            database.setString(
+                conversationId + "/" + KEYS_SUB_PATH + "/" + currentMessenger.name,
+                encodedPublicKey
+            )
+        }
+
+    }
 
     /**
      * Sends a message from the current user to the database
      * @param messageText The message text
      */
     fun sendMessage(messageText: String) {
-        val message = Message(currentMessenger, messageText)
-        database.addToObjectsListConcurrently(conversationId, Message::class.java, message)
+        // Retrieve public key from Bob
+        publicKey?.let { publicKey ->
+            encryptAndSend(publicKey!!, messageText)
+        }?.run {
+            val path = conversationId + "/" + KEYS_SUB_PATH + "/" +
+                    Messenger.values()[(currentMessenger.ordinal + 1) % 2].name
+
+            database.getString(path).thenApply {
+                val decodedKey = Base64.decode(it, Base64.DEFAULT)
+                val keySpecs = X509EncodedKeySpec(decodedKey)
+                publicKey = KeyFactory.getInstance("RSA").generatePublic(keySpecs)
+                encryptAndSend(publicKey!!, messageText)
+            }
+        }
+    }
+
+    /**
+     * Encrypts the given message with the given public key and sends it to the
+     * database
+     * @param publicKey The public key to be used for encryption
+     * @param message The message to encrypt
+     */
+    private fun encryptAndSend(publicKey: PublicKey, message: String) {
+        val cipher = Cipher.getInstance(TRANSFORMATION)
+        cipher.init(Cipher.ENCRYPT_MODE, publicKey)
+
+        val bytes = cipher.doFinal(message.toByteArray())
+
+        val encryptedMessage =
+            Message(currentMessenger, Base64.encodeToString(bytes, Base64.DEFAULT))
+
+        allMessages[encryptedMessage.message] = message
```

Comment:
This mapping get deleted when the activity closes => It should be recomputed every time?
You should definitely put it in a local storage (not connected to internet, obviously ;))